### DISCUSSION TITLE
automatic rekeying issue

### DISCUSSION THREAD

**Discussion by SB8767** [strongswan/strongswan#2318] [Q&A]
I have a strongswan VPN that does automatically rekey and I have to manually run 'ipsec up vpn-(customername) to re-establish connection.

I tried adding
`rekey=yes`
`keyingtries=%forever`
`rekeymargin=3m`

I also changed the lifetime from 28800s to 3600s just so I can see the logs. The VPN dropped after 50 minutes or so and this is what I found in the logs:

grep -rn /var/log/syslog -e "charon" 

- 1300452:Jul  1 11:26:36 admin-1 charon: 10[IKE] received DELETE for ESP CHILD_SA with SPI f7e8dcb0
- 1300453:Jul  1 11:26:36 admin-1 charon: 10[IKE] closing CHILD_SA vpn-*{3} with SPIs c63f2ce7_i (43384 bytes) f7e8dcb0_o (54032 bytes) and TS 10.241.242.40/29 === 10.240.209.120/29
- 1300454:Jul  1 11:26:36 admin-1 charon: 10[IKE] sending DELETE for ESP CHILD_SA with SPI c63f2ce7
- 1300455:Jul  1 11:26:36 admin-1 charon: 10[IKE] CHILD_SA closed
- 1300457:Jul  1 11:26:36 admin-1 charon: 10[IKE] detected CHILD_REKEY collision with CHILD_DELETE
- 1300458:Jul  1 11:26:36 admin-1 charon: 10[ENC] generating INFORMATIONAL response 147 [ D ]

What can I add to my ipsec.conf to ensure the ike_sa can automatic rekey?

**Comment by tobiasbrunner** [comment]
The log is a bit short, but it's possible that your peer does rekeying incorrectly, i.e. it sends a delete for the old CHILD_SA as a responder of a rekeying, which it shouldn't do. Either use the latest 6.0 pre-release (or the _multi-ke_ branch), which can handle this incorrect behavior, or let the peer do the rekeying.

**Reply by SB8767** [reply]
Thank you.

Do you know where I can find more information on multi-ke branch? I found the link for the update: https://www.strongswan.org/blog/2022/07/29/strongswan-5.9.7-released.html but I would like to understand more about the multi-ke branch?

**Reply by tobiasbrunner** [reply]
That's not what I was referring to. The [multi-ke branch](https://github.com/strongswan/strongswan/compare/master...multi-ke) has not yet been merged (although, I might actually do so soon).

**Reply by SB8767** [reply]
I see, I apologise. So it's either update strongswan to 6.0 or find out what timeout settings are set on the peer side?

**Reply by tobiasbrunner** [reply]
You could also disable rekeying, it will then be up to the peer if the SAs are rekeyed or not.

**Reply by SB8767** [reply]
The customer says they have a script that if the VPN goes down, the script automatically brings it back up. The customer also said this is the only vpn where the re-keying does not automatically work.


Also, I added the below to ipsec.conf today.
rekey=yes
keyingtries=%forever
rekeymargin=3m

So, in theory, the VPN should have rekeyed last week but it did not.

**Reply by tobiasbrunner** [reply]
> The customer says they have a script that if the VPN goes down, the script automatically brings it back up.

How does it go down during a rekeying in a way that would trigger their script? The log excerpt (again, it's a bit short to say for sure) seems to show a CHILD_SA rekeying. Nothing is going down during that. In particular not the IKE_SA.

> So, in theory, the VPN should have rekeyed last week but it did not.

What do you mean?

**Reply by SB8767** [reply]
> How does it go down during a rekeying in a way that would trigger their script? The log excerpt (again, it's a bit short to say for sure) seems to show a CHILD_SA rekeying. Nothing is going down during that. In particular not the IKE_SA.

Sorry, I was mistaken. I spoke to the customer again and the script isn't used to bring the VPN up, so please ignore my previous message.

I asked the customer "what happens if the VPN goes down? How is rekeying configured on your end?" The customer said the VPN automatically rekeys. The customer went on to say it works with other customers, and it is only an issue with us, and we're the only ones using StrongSwan.

So I have to update StrongSwan to 6.0 to fix the issue?

**Reply by tobiasbrunner** [reply]
If they have rekeying enabled, then you can disable it from your end so rekeying is only initiated by the peer. Or you can try 6.0, which should also work if rekeying is initiated from your end.

**Reply by SB8767** [reply]
Ah I see. To answer your earlier question

> So, in theory, the VPN should have rekeyed last week but it did not.
> What do you mean?

Last week I did not have rekeying enabled on my side but the customer had rekeying enabled. The problem is, the rekeying did not work. The VPN would drop after an hour (this was defined by lifetime=3600s and I will change this once I solve this issue).

**Reply by tobiasbrunner** [reply]
Read the logs for why the rekeiyng failed when it was initiated by the peer.

**Reply by SB8767** [reply]
I shortened the lifetime so I can generate logs and I also removed the rekey entry in ipsec.conf. 

See below ipsec.conf and logs:


conn vpn-test
    type=tunnel
    auto=start
    keyexchange=ikev2
    authby=secret
    left= myip
    leftsubnet= 10.241.242.40/29
    leftfirewall=yes
    right= customerip
    rightsubnet=10.240.209.120/29
    rightfirewall=yes
    ike=aes256-sha512-modp2048,aes256-sha384-modp2048,aes256-sha256-modp2048,aes256-sha-modp2048!
    esp=aes256-sha512-modp2048,aes256-sha384-modp2048,aes256-sha256-modp2048,aes256-sha1-modp2048!
    dpdaction=restart
    ikelifetime=86400s
    lifetime=1800s
    dpddelay=20s
    dpdtimeout=60s
    forceencaps=yes




logs:

1853895:Jul  2 15:51:51 admin charon: 13[IKE] peer didn't accept DH group MODP_2048, it requested MODP_NONE
1853896:Jul  2 15:51:51 admin charon: 13[IKE] establishing CHILD_SA vpn-vpn-*customername*{12} reqid 4
1853897:Jul  2 15:51:51 admin charon: 13[ENC] generating CREATE_CHILD_SA request 6 [ N(REKEY_SA) SA No TSi TSr ]
1853900:Jul  2 15:51:51 admin charon: 09[ENC] parsed CREATE_CHILD_SA response 6 [ SA No TSi TSr ]
1853901:Jul  2 15:51:51 admin charon: 09[IKE] no acceptable proposal found
1853902:Jul  2 15:51:51 admin charon: 09[IKE] failed to establish CHILD_SA, keeping IKE_SA
1853903:Jul  2 15:51:51 admin charon: 09[IKE] sending DELETE for ESP CHILD_SA with SPI cab32cff
1853904:Jul  2 15:51:51 admin charon: 09[ENC] generating INFORMATIONAL request 7 [ D ]
1853905:Jul  2 15:51:51 admin charon: 09[NET] sending packet: from myip [4500] to customerip [4500] (96 bytes)
1853906:Jul  2 15:51:51 admin charon: 07[NET] received packet: from customerip[4500] to myip[4500] (96 bytes)
1853907:Jul  2 15:51:51 admin charon: 07[ENC] parsed INFORMATIONAL request 36 [ D ]
1853908:Jul  2 15:51:51 admin charon: 07[IKE] received DELETE for ESP CHILD_SA with SPI c4bb942c
1853909:Jul  2 15:51:51 admin charon: 07[IKE] closing CHILD_SA vpn-vpn-*customername* {8} with SPIs c1009220_i (9860 bytes) c4bb942c_o (12320 bytes) and TS 10.241.242.40/29 === 10.240.209.120/29
1853910:Jul  2 15:51:51 admin charon: 07[IKE] sending DELETE for ESP CHILD_SA with SPI c1009220
1853911:Jul  2 15:51:51 admin charon: 07[IKE] CHILD_SA closed
1853913:Jul  2 15:51:51 admin charon: 07[IKE] detected CHILD_REKEY collision with CHILD_DELETE
1853914:Jul  2 15:51:51 admin charon: 07[ENC] generating INFORMATIONAL response 36 [ D ]
1853915:Jul  2 15:51:51 admin charon: 07[NET] sending packet: from 161.35.46.168[4500] to customerip [4500] (96 bytes)
1853916:Jul  2 15:51:51 admin charon: 16[NET] received packet: from customerip[4500] to myip[4500] (96 bytes)
1853917:Jul  2 15:51:51 admin charon: 16[ENC] parsed INFORMATIONAL response 7 [ ]
1853918:Jul  2 15:51:51 admin charon: 14[NET] received packet: from customerip[4500] to myip [4500] (96 bytes)

**Reply by tobiasbrunner** [reply]
> 1853895:Jul 2 15:51:51 admin charon: 13[IKE] peer didn't accept DH group MODP_2048, it requested MODP_NONE
> ...
> 1853901:Jul 2 15:51:51 admin charon: 09[IKE] no acceptable proposal found
> 1853902:Jul 2 15:51:51 admin charon: 09[IKE] failed to establish CHILD_SA, keeping IKE_SA

The rekeying fails because of a misconfiguration of your proposals. The peer is not configured to do a DH exchange during CHILD_SA rekeying (PFS). However, your ESP proposals all list `modp2048` as DH group. So either remove that group, or add `modpnone` to all the proposals them do make the DH group optional (after the existing DH group, like `modp2048-modpnone`). Or you ask your peer to enable PFS.
